File: dmarc_dashboard/fetcher.py
"""Fetch DMARC reports from Microsoft 365 and persist them to the database."""
import logging
from typing import Dict, Any

from dmarc_dashboard import db
from dmarc_dashboard.graph_client import GraphClient
from dmarc_dashboard.dmarc_parser import parse_report
from dmarc_dashboard.models import Report, Record

logger = logging.getLogger(__name__)


def fetch_and_store(config: Dict[str, Any]) -> int:
    client = GraphClient(
        tenant_id=config["TENANT_ID"],
        client_id=config["CLIENT_ID"],
        client_secret=config["CLIENT_SECRET"],
    )
    mailbox = config["MAILBOX"]
    folder = config.get("MAIL_FOLDER", "Inbox")

    messages = client.get_dmarc_messages(mailbox, folder)
    stored = 0

    for msg in messages:
        msg_id = msg["id"]
        if Report.query.filter_by(email_message_id=msg_id).first():
            continue

        xml_payloads = client.get_xml_attachments(mailbox, msg_id)
        for xml_bytes in xml_payloads:
            data = parse_report(xml_bytes)
            if not data:
                logger.warning("Failed to parse XML from message %s", msg_id)
                continue

            if Report.query.filter_by(report_id=data["report_id"]).first():
                logger.debug("Report %s already in DB, skipping", data["report_id"])
                continue

            report = Report(
                report_id=data["report_id"],
                org_name=data["org_name"],
                org_email=data["org_email"],
                domain=data["domain"],
                begin_date=data["begin_date"],
                end_date=data["end_date"],
                policy_p=data["policy_p"],
                policy_sp=data["policy_sp"],
                policy_pct=data["policy_pct"],
                adkim=data["adkim"],
                aspf=data["aspf"],
                email_message_id=msg_id,
            )
            db.session.add(report)
            db.session.flush()

            for rec in data["records"]:
                db.session.add(Record(report_id=report.id, **rec))

            db.session.commit()
            stored += 1
            logger.info("Stored report %s from %s", data["report_id"], data["org_name"])

    return stored

File: src/fetcher.js
const { getDb } = require('./db');
const { GraphClient } = require('./graphClient');
const { parseReport } = require('./dmarcParser');

async function fetchAndStore(tenant, accessToken = null) {
  const client = new GraphClient(
    accessToken
      ? { accessToken }
      : { tenantId: tenant.tenant_id, clientId: tenant.client_id, clientSecret: tenant.client_secret }
  );
  const db = getDb();
  const insertReport = db.prepare(`
    INSERT INTO reports
      (report_id, org_name, org_email, domain, begin_date, end_date,
       policy_p, policy_sp, policy_pct, adkim, aspf, email_message_id, tenant_db_id)
    VALUES
      (@report_id, @org_name, @org_email, @domain, @begin_date, @end_date,
       @policy_p, @policy_sp, @policy_pct, @adkim, @aspf, @email_message_id, @tenant_db_id)
  `);
  const insertRecord = db.prepare(`
    INSERT INTO records
      (report_id, source_ip, count, disposition, dkim_aligned, spf_aligned,
       header_from, envelope_from, dkim_domain, dkim_selector, dkim_auth_result,
       spf_domain, spf_auth_result)
    VALUES
      (@report_id, @source_ip, @count, @disposition, @dkim_aligned, @spf_aligned,
       @header_from, @envelope_from, @dkim_domain, @dkim_selector, @dkim_auth_result,
       @spf_domain, @spf_auth_result)
  `);

  const messages = await client.getDmarcMessages(tenant.mailbox, tenant.mail_folder);
  let stored = 0;

  for (const msg of messages) {
    const msgId = msg.id;
    if (db.prepare('SELECT id FROM reports WHERE email_message_id = ?').get(msgId)) continue;

    const xmlPayloads = await client.getXmlAttachments(tenant.mailbox, msgId);
    for (const xml of xmlPayloads) {
      const data = parseReport(xml);
      if (!data) { console.warn(`Failed to parse XML from message ${msgId}`); continue; }
      if (db.prepare('SELECT id FROM reports WHERE report_id = ?').get(data.report_id)) continue;

      db.transaction(() => {
        const { lastInsertRowid } = insertReport.run({ ...data, email_message_id: msgId, tenant_db_id: tenant.id });
        for (const rec of data.records) insertRecord.run({ ...rec, report_id: lastInsertRowid });
      })();
      stored++;
      console.log(`[fetch] Stored report ${data.report_id} from ${data.org_name} (tenant: ${tenant.name})`);
    }
  }
  return stored;
}

module.exports = { fetchAndStore };
